server/src/main/java/com/thoughtworks/go/server/service/ConfigRepositoryInitializer.java

```diff
@@ -0,0 +1,146 @@
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.server.service;
+
+import com.thoughtworks.go.config.CruiseConfig;
+import com.thoughtworks.go.config.GoConfigRepoConfigDataSource;
+import com.thoughtworks.go.config.materials.MaterialConfigs;
+import com.thoughtworks.go.config.materials.Materials;
+import com.thoughtworks.go.config.remote.ConfigRepoConfig;
+import com.thoughtworks.go.domain.MaterialInstance;
+import com.thoughtworks.go.domain.MaterialRevisions;
+import com.thoughtworks.go.domain.materials.Material;
+import com.thoughtworks.go.domain.materials.MaterialConfig;
+import com.thoughtworks.go.domain.materials.Modification;
+import com.thoughtworks.go.listener.ConfigChangedListener;
+import com.thoughtworks.go.plugin.infra.PluginChangeListener;
+import com.thoughtworks.go.plugin.infra.PluginManager;
+import com.thoughtworks.go.plugin.infra.plugininfo.GoPluginDescriptor;
+import com.thoughtworks.go.server.initializers.Initializer;
+import com.thoughtworks.go.server.persistence.MaterialRepository;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.io.File;
+import java.util.LinkedList;
+
+import static com.thoughtworks.go.plugin.domain.common.PluginConstants.CONFIG_REPO_EXTENSION;
+
+/**
+ * @understands initializing config repositories. Loads the configurations from the last checked out modification on server startup.
+ */
+@Service
+public class ConfigRepositoryInitializer implements ConfigChangedListener, PluginChangeListener, Initializer {
+    private PluginManager pluginManager;
+    private final ConfigRepoService configRepoService;
+    private final MaterialRepository materialRepository;
+    private final GoConfigRepoConfigDataSource goConfigRepoConfigDataSource;
+    private GoConfigService goConfigService;
+    private static final Logger LOGGER = LoggerFactory.getLogger(ConfigRepositoryInitializer.class);
+
+    private boolean isConfigLoaded = false;
+
+    // list of config repo plugins which are loaded, but not yet processed. Once processed, these plugins will be removed from the list.
+    private final LinkedList<String> pluginsQueue = new LinkedList<>();
+
+    @Autowired
+    public ConfigRepositoryInitializer(PluginManager pluginManager, ConfigRepoService configRepoService, MaterialRepository materialRepository, GoConfigRepoConfigDataSource goConfigRepoConfigDataSource, GoConfigService goConfigService) {
+        this.pluginManager = pluginManager;
+        this.configRepoService = configRepoService;
+        this.materialRepository = materialRepository;
+        this.goConfigRepoConfigDataSource = goConfigRepoConfigDataSource;
+        this.goConfigService = goConfigService;
+
+        this.pluginManager.addPluginChangeListener(this);
+    }
+
+    @Override
+    public void initialize() {
+        this.goConfigService.register(this);
```

Comment:
@GaneshSPatil can we register for ConfigChange in the constructor over the `intitalize` method. If we register in the `intialize` method, the config repos will most likely be parsed as part of the main thread and this will lead to increased startup time for GoCD instances with larger configs and multiple config repos. WDYT?